File: Practice_2 (Java)/src/CommandsHandler.java
import java.util.Iterator;
import java.util.Scanner;
import ua.KarpovIvan.atd.*;

public class CommandsHandler {
    static Scanner in = new Scanner(System.in);

    /*
     * Команда для виводу списока всіх доступних команд.
     */
    public static void commandHelp() {
        System.out.println(" Доступнi команди:");
        System.out.println(" /addfirst\t- додати елемент в початок списку.");
        System.out.println(" /clear\t\t- очистити консоль.");
        System.out.println(" /end\t\t- закiнчити роботу.");
        System.out.println(" /help\t\t- вивести список доступних команд.");
        System.out.println(" /length\t- вивести довжину списку.");
        System.out.println(" /removelast\t- видалити елемент з кiнця списку.");
        System.out.println(" /search\t- знайти у списку студентiв з iдеальною вагою.");
        System.out.println(" /show\t\t- вивести список.");
        System.out.println(" /showback\t- вивести список у зворотньому порядку.");
        System.out.println(" /sort\t\t- сортувати список зростом.");
    }

    /*
     * Команда для додавання елемента в початок списку.
     */
    public static void commandAddFirst(DoublyLinkedList<Student> students) {
        /* Ввід імені студента з перевіркою на правильність та допустимість діапазону. */
        Names name;
        System.out.println(" Введiть iм'я студента");
        System.out.println(" (0 - Андрiй, " +
                "1 - Вiка, " +
                "2 - Денис, " +
                "3 - iван, " +
                "4 - iнна, " +
                "5 - Лiза, " +
                "6 - Микола, " +
                "7 - Мирослав, " +
                "8 - Сергiй, " +
                "9 - Снiжана):");
        while (true) {
            try {
                System.out.print(" ");
                int index = in.nextInt(); /* Зчитування числа. */
                if (-1 < index && index < Names.Count.ordinal()) {
                    name = Names.values()[index];
                    break;
                }
                else {
                    throw new Exception();
                }
            }
            catch (Exception exception) {
                System.out.println(" Введено невiрне значення! Спробуйте ще раз:");
                in.nextLine();
            }
        }

        /* Ввід зросту студента з перевіркою на правильність та допустимість діапазону. */
        int height;
        System.out.println(" Введiть зрiст студента в межах вiд 50 см до 250 см (цiле значення):");
        while (true) {
            try {
                System.out.print(" ");
                height = in.nextInt();

                if (50 <= height && height <= 250) {
                    break;
                }
                else {
                    throw new Exception();
                }
            }
            catch (Exception exception) {
                System.out.println(" Введено невiрне значення! Спробуйте ще раз:");
                in.nextLine();
            }
        }

        /* Ввід ваги студента з перевіркою на правильність та допустимість діапазону. */
        double weight;
        System.out.println(" Введiть вагу студента вiд 40 кг до 250 кг:");
        while (true) {
            try {
                System.out.print(" ");
                weight = in.nextDouble();

                if (40 <= weight && weight <= 250) {
                    break;
                }
                else {
                    throw new Exception();
                }
            }
            catch (Exception exception) {
                System.out.println(" Введено невiрне значення! Спробуйте ще раз:");
                in.nextLine();
            }
        }

        /* Створення нового екземпляру Student да додавання його до списку. */
        Student student = new Student(name, height, weight);
        students.addFirst(student);

        /* Виведення оновленого списоку на екран. */
        commandShow(students);
    }

    /*
     * Команда для виводу довжини списку.
     */
    public static void commandLength(DoublyLinkedList<Student> students) {
        System.out.printf(" Довжина списку: %d\n", students.getCount());
    }

    /*
     * Команда для видалення останнього елементу зі списку.
     */
    public static void commandRemoveLast(DoublyLinkedList<Student> students) {
        try {
            students.removeLast();
        }
        catch (Exception exception) {
            System.out.printf(" %s\n", exception.getMessage());
        }

        /* Виводить оновлений список на екран. */
        commandShow(students);
    }

    /*
     * Команда для пошуку і виводу на консоль студента з ідеальною вагою тіла.
     */
    public static void commandSearch(DoublyLinkedList<Student> students) {
        try {
            DoublyLinkedList<Student> searchResult = students.searchOfElement();
            System.out.println(" Результати пошуку:");
            commandShow(searchResult);
        }
        catch (Exception exception) {
            System.out.printf(" %s\n", exception.getMessage());
        }
    }

    /*
     * Команда для виводу списку на консоль у вигляді таблиці.
     */
    public static void commandShow(DoublyLinkedList<Student> students) {
        String hyphensLine = String.format("*%38s", "-").replace(' ', '-').replace('*', ' ');
        System.out.println(hyphensLine);
        System.out.printf(" |  %8s  |  %8s  |  %6s  |%n", "Iм'я", "Зрiст", "Вага");
        System.out.println(hyphensLine);

        /* Виведення порожньої таблиці або таблиці з елементами списку. */
        if (students.listEmpty()) {
            System.out.printf(" |  %8s  |  %8s  |  %6s  |%n", "", "", "");
        }
        else {
            for (Student student : students) {
                System.out.printf(" |  %8s  |  %8d  |  %6.2f  |%n", student.getName().name(), student.getHeight(), student.getWeight());
            }
        }
        System.out.println(hyphensLine);
    }

    /*
     * Команда для виводу списку на консоль у вигляді таблиці у зворотному порядку.
     */
    public static void commandShowBack(DoublyLinkedList<Student> students) {
        String hyphensLine = String.format("*%38s", "-").replace(' ', '-').replace('*', ' ');
        System.out.println(hyphensLine);
        System.out.printf(" |  %8s  |  %8s  |  %6s  |%n", "Iм'я", "Зрiст", "Вага");
        System.out.println(hyphensLine);

        /* Виведення порожньої таблиці або таблиці з елементами списку. */
        if (students.listEmpty()) {
            System.out.printf(" |  %8s  |  %8s  |  %6s  |%n", "", "", "");
        }
        else {
            Iterator<Student> iterator = students.backIterator();

            while (iterator.hasNext()) {
                Student student = iterator.next();
                System.out.printf(" |  %8s  |  %8d  |  %6.2f  |%n", student.getName().name(), student.getHeight(), student.getWeight());
            }
        }
        System.out.println(hyphensLine);
    }

    /*
     * Командля для сортування списку.
     */
    public static void commandSort(DoublyLinkedList<Student> students) {
        try {
            students.sort();
            commandShow(students);
        }
        catch (Exception exception) {
            System.out.printf(" %s\n", exception.getMessage());
        }
    }
}  

File: Practice_1 (C#)/Practice_1/CommandsHandler.cs
using System.Collections.Generic;
using System.Collections;
using System;
using ListLib;

namespace Practice_1
{
    internal class CommandsHandler
    {
        // Команда для виводу списока всіх доступних команд.
        public static void CommandHelp()
        {
            Console.WriteLine(" Доступні команди:");
            Console.WriteLine(" /addfirst\t- додати елемент в початок списку.");
            Console.WriteLine(" /clear\t\t- очистити консоль.");
            Console.WriteLine(" /end\t\t- закінчити роботу.");
            Console.WriteLine(" /help\t\t- вивести список доступних команд.");
            Console.WriteLine(" /length\t- вивести довжину списку.");
            Console.WriteLine(" /removelast\t- видалити елемент з кінця списку.");
            Console.WriteLine(" /search\t- знайти у списку студентів відмінників, що займаються спортом.");
            Console.WriteLine(" /show\t\t- вивести список.");
            Console.WriteLine(" /showback\t- вивести список у зворотньому порядку.");
            Console.WriteLine(" /sort\t\t- сортувати список за середнім балом.");
        }

        // Команда для додавання елемента в початок списку.
        public static void CommandAddFirst(ref DoublyLinkedList<Student> students)
        {
            // Ввід імені студента з перевіркою на правильність.
            Names name;
            Console.WriteLine(" Введіть ім'я студента");
            Console.WriteLine(" (0 - Андрій, " +
                                "1 - Віка, " +
                                "2 - Денис, " +
                                "3 - Іван, " +
                                "4 - Інна, " +
                                "5 - Ліза, " +
                                "6 - Микола, " +
                                "7 - Мирослав, " +
                                "8 - Сергій, " +
                                "9 - Сніжана):");  
            while (true)
            {
                // Використання блоку try/catch потрібне, щоб запобігти випадку,
                // коли рядок типу будь-чого, окрім числових значень, пробує конвертуватись в enum.
                try
                {
                    Console.Write(" ");
                    name = (Names)Convert.ToInt16(Console.ReadLine());
                    if (0 <= name && name < Names.Count)
                    {
                        break;
                    }
                    else
                    {
                        throw new Exception();
                    }
                }
                catch
                {
                    Console.WriteLine(" Введено невірне значення! Спробуйте ще раз:");
                }                
            }

            // Ввід середнього балу студента з перевіркою на правильність.
            double gpa;
            Console.WriteLine(" Введіть середній бал (від 0 до 100 в форматі XX,X):");
            while (true)
            {
                // Використання блоку try/catch потрібне, щоб запобігти випадку,
                // коли рядок типу "XX.X" пробує конвертуватись в дійсний тип.
                // Дана дія визиває виключення FormatException.
                try
                {
                    Console.Write(" ");
                    gpa = Convert.ToDouble(Console.ReadLine());

                    if (0 <= gpa && gpa <= 100)
                    {
                        break;
                    }
                    else
                    {
                        throw new Exception();
                    }
                }
                catch
                {
                    Console.WriteLine(" Введено невірне значення! Спробуйте ще раз:");
                }                
            }

            // Ввід логічно значення залежного від того чи займається студент спортом, чи ні,
            // з перевіркою на правильність.
            bool sport;
            Console.WriteLine(" Виберіть чи займається студент спортом (Y/N):");
            while (true)
            {
                Console.Write(" ");
                string answer = Console.ReadLine().ToLower();
                if (answer == "y" || answer == "n")
                {
                    sport = answer == "y" ? true : false;
                    break;
                }
                else
                {
                    Console.WriteLine(" Введено невірне значення! Спробуйте ще раз:");
                }
            }

            // Створення нового екземпляру Student да додавання його до списку.
            var student = new Student(name, gpa, sport);
            students.AddFirst(student);

            // Виводить оновлений список на екран.
            CommandShow(ref students);
        }

        // Команда для виводу довжини списку.
        public static void CommandLenght(ref DoublyLinkedList<Student> students)
        {
            Console.WriteLine(" Довжина списку: {0}", students.Count);
        }

        // Команда для видалення останнього елементу зі списку.
        public static void CommandRemoveLast(ref DoublyLinkedList<Student> students)
        {
            try
            {
                students.RemoveLast();
            }
            catch (Exception exception)
            {
                Console.WriteLine($" {exception.Message}");
            }

            // Виводить оновлений список на екран.
            CommandShow(ref students);
        }

        // Команда для пошуку і виводу на консоль студента відмінника, який займається спортом.
        public static void CommandSearch(ref DoublyLinkedList<Student> students)
        {
            try
            {
                DoublyLinkedList<Student> searchResult = students.SearchOfElement();
                Console.WriteLine(" Результати пошуку:");
                CommandShow(ref searchResult);
            }
            catch (Exception exception)
            {
                Console.WriteLine($" {exception.Message}");
            }            
        }

        // Команда для виводу списку на консоль у вигляді таблиці.
        public static void CommandShow(ref DoublyLinkedList<Student> students)
        {
            Console.Write("-".PadLeft(42, '-'));
            Console.Write("\r \n");
            Console.WriteLine(" |  {0,8}  |  {1,5}  |  {2,5}  |", "Ім'я", "Середній бал", "Спорт");
            Console.Write("-".PadLeft(42, '-'));
            Console.Write("\r \n");

            if (students.ListEmpty)
            {
                Console.WriteLine(" |  {0,8}  |  {1,12}  |  {2,5}  |", "", "", "");
            }
            else
            {
                foreach (var student in students)
                {
                    Console.WriteLine(" |  {0,8}  |  {1,12}  |  {2,5}  |", student.Name, student.Gpa, student.Sport);
                }
            }            
            Console.Write("-".PadLeft(42, '-'));
            Console.Write("\r \n");
        }

        // Команда для виводу списку на консоль у вигляді таблиці у зворотному порядку.
        public static void CommandShowBack(ref DoublyLinkedList<Student> students)
        {
            Console.Write("-".PadLeft(42, '-'));
            Console.Write("\r \n");
            Console.WriteLine(" |  {0,8}  |  {1,5}  |  {2,5}  |", "Ім'я", "Середній бал", "Спорт");
            Console.Write("-".PadLeft(42, '-'));
            Console.Write("\r \n");

            if (students.ListEmpty)
            {
                Console.WriteLine(" |  {0,8}  |  {1,12}  |  {2,5}  |", "", "", "");
            }
            else
            {
                foreach (var student in students.BackEnumerator())
                {
                    Console.WriteLine(" |  {0,8}  |  {1,12}  |  {2,5}  |", student.Name, student.Gpa, student.Sport);
                }
            }
            Console.Write("-".PadLeft(42, '-'));
            Console.Write("\r \n");
        }

        // Командля для сортування списку.
        public static void CommandSort(ref DoublyLinkedList<Student> students)
        {
            try
            {
                students.Sort();
            }
            catch (Exception exception)
            {
                Console.WriteLine($" {exception.Message}");
            }
            
            CommandShow(ref students);
        }
    }    
}
